File: composeApp/src/androidMain/kotlin/org/gce/racehub/signup/SignUpViewModel.kt
package org.gce.racehub.signup

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.gce.racehub.auth.domain.model.OtpPurpose
import org.gce.racehub.auth.domain.usecase.SendOtpUseCase
import org.gce.racehub.auth.domain.usecase.SignUpUseCase

class SignUpViewModel(
    private val signUpUseCase: SignUpUseCase,
    private val sendOtpUseCase: SendOtpUseCase
) : ViewModel() {

    private val _state = MutableStateFlow(SignUpState())
    val state: StateFlow<SignUpState> = _state.asStateFlow()

    private val _effect = Channel<SignUpEffect>(Channel.BUFFERED)
    val effect = _effect.receiveAsFlow()

    fun onIntent(intent: SignUpIntent) {
        when (intent) {
            is SignUpIntent.UsernameChanged ->
                _state.update { it.copy(username = intent.username) }

            is SignUpIntent.EmailChanged ->
                _state.update { it.copy(email = intent.email) }

            is SignUpIntent.PasswordChanged ->
                _state.update { it.copy(password = intent.password) }

            is SignUpIntent.ConfirmPasswordChanged ->
                _state.update { it.copy(confirmPassword = intent.confirmPassword) }

            is SignUpIntent.CountryChanged ->
                _state.update { it.copy(country = intent.country) }

            is SignUpIntent.TogglePasswordVisibility ->
                _state.update { it.copy(isPasswordVisible = !it.isPasswordVisible) }

            is SignUpIntent.ToggleConfirmPasswordVisibility ->
                _state.update { it.copy(isConfirmPasswordVisible = !it.isConfirmPasswordVisible) }

            is SignUpIntent.SignUp ->
                performSignUp()

            is SignUpIntent.DismissError ->
                _state.update { it.copy(errorMessage = null) }
        }
    }

    private fun performSignUp() {
        if (_state.value.isLoading) return
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, errorMessage = null) }

            val s = _state.value
            val result = signUpUseCase(s.username, s.email, s.password, s.country, s.confirmPassword)

            if (!result.isSuccess) {
                _state.update { it.copy(isLoading = false, errorMessage = result.error) }
                return@launch
            }

            // Account created but NOT logged in: the user must verify their email
            // via OTP before the sign-up is considered complete.
            val otpResult = sendOtpUseCase(s.email, OtpPurpose.EMAIL_VERIFICATION)
            if (otpResult.isSuccess) {
                _state.update { it.copy(isLoading = false) }
                _effect.send(SignUpEffect.NavigateToEmailVerification(s.email))
            } else {
                _state.update { it.copy(isLoading = false, errorMessage = otpResult.error) }
            }
        }
    }
}

File: iosApp/iosApp/SignUp/SignUpViewModel.swift
import Foundation
import Combine
import Shared

/// ViewModel for the Sign-Up screen following the MVI pattern.
///
/// - Exposes `state` as a `@Published` property the View observes.
/// - Accepts user actions via `send(_:)` and processes them into state mutations.
/// - Emits one-time navigation events through `effectPublisher`.
///
/// `@MainActor` ensures all state mutations happen on the main thread,
/// keeping SwiftUI updates safe without manual `DispatchQueue.main` calls.
@MainActor
final class SignUpViewModel: ObservableObject {

    /// Observable UI state. Collected by the View via `@StateObject` + `viewModel.state`.
    @Published private(set) var state = SignUpState()

    // PassthroughSubject is used so each effect fires exactly once and is not
    // replayed to late subscribers (unlike CurrentValueSubject).
    private let effectSubject = PassthroughSubject<SignUpEffect, Never>()

    /// Stream of one-time side effects (navigation, etc.) for the View to react to.
    var effectPublisher: AnyPublisher<SignUpEffect, Never> {
        effectSubject.eraseToAnyPublisher()
    }

    /// Use case that validates the sign-up form and delegates to the repository.
    private let signUpUseCase: SignUpUseCase

    init() {
        signUpUseCase = SignUpUseCase(authRepository: AuthRepositoryImpl())
    }

    /// Entry point for all View interactions.
    /// Maps each `SignUpIntent` to a state mutation or a command.
    func send(_ intent: SignUpIntent) {
        switch intent {
        case .nameChanged(let name):
            state.name = name

        case .emailChanged(let email):
            state.email = email

        case .passwordChanged(let password):
            state.password = password

        case .confirmPasswordChanged(let confirmPassword):
            state.confirmPassword = confirmPassword

        case .togglePasswordVisibility:
            state.isPasswordVisible.toggle()

        case .toggleConfirmPasswordVisibility:
            state.isConfirmPasswordVisible.toggle()

        case .signUp:
            performSignUp()
        }
    }

    /// Runs the sign-up use case and updates state based on the result.
    /// Wrapped in a `Task` so the synchronous use case call does not block the
    /// main thread when replaced with a real async implementation.
    private func performSignUp() {
        state.isLoading = true
        state.errorMessage = nil

        Task {
            let result = signUpUseCase.execute(
                name: state.name,
                email: state.email,
                password: state.password,
                confirmPassword: state.confirmPassword
            )
            state.isLoading = false

            if result.isSuccess {
                effectSubject.send(.navigateToHome)
            } else {
                state.errorMessage = result.error
            }
        }
    }
}
